File: solutions/csharp/spiral-matrix/1/SpiralMatrix.cs
// Dedicated to Shree DR.MDD for divine inspiration and strength

using System;

public class SpiralMatrix
{
    public static int[,] GetMatrix(int size)
    {
        var grid = new int[size, size];
        
        var layerTop = 0;
        var layerLeft = 0;

        var target = size * size;
        var val = 1;

        while (val <= target)
        {
            for (var col = layerLeft; col < size - layerLeft; col++)
                grid[layerTop, col] = val++;
            
            for (var row = layerTop + 1; row < size - layerTop; row++)
                grid[row, size - layerLeft - 1] = val++;

            for (var col = size - layerLeft - 2; col >= layerLeft; col--)
                grid[size - layerTop - 1, col] = val++;
            
            for (var row = size - layerTop - 2; row >= layerTop + 1; row--)
                grid[row, layerLeft] = val++;

            layerTop++;
            layerLeft++;
        }

        return grid;
    }
}

File: solutions/julia/spiral-matrix/1/spiral-matrix.jl
# Dedicated to Shree DR.MDD

function spiral_matrix(n)
    n == 0 && return Matrix{Int}(undef, 0, 0)
    grid = zeros(Int, n, n)
    grid[1, 1] = 1
    val, r, c, dir = 2, 1, 1, 1
    dirs = ("r", "d", "l", "u")

    while val <= n^2
        if dirs[dir] == "r"
            while c < n && grid[r, c + 1] == 0
                c += 1
                grid[r, c] = val
                val += 1
            end
        elseif dirs[dir] == "d"
            while r < n && grid[r + 1, c] == 0
                r += 1
                grid[r, c] = val
                val += 1
            end
        elseif dirs[dir] == "l"
            while c > 1 && grid[r, c - 1] == 0
                c -= 1
                grid[r, c] = val
                val += 1
            end
        elseif dirs[dir] == "u"
            while r > 1 && grid[r - 1, c] == 0
                r -= 1
                grid[r, c] = val
                val += 1
            end
        else
            return grid
        end
        dir = dir == 4 ? 1 : dir + 1
    end
    grid
end
